File: LeetCode/JavaScript/0019-Remove-Nth-Node-From-End-of-List.js
// 19. Remove Nth Node From End of List
// https://leetcode.com/problems/remove-nth-node-from-end-of-list/

/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */

/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */

var removeNthFromEnd = function (head, n) {
  const startNode = new ListNode(0);
  startNode.next = head;

  let prev = (cur = startNode);
  for (let i = 0; i <= n; i++) cur = cur.next;

  while (cur != null) {
    cur = cur.next;
    prev = prev.next;
  }
  prev.next = prev.next.next;
  return startNode.next;
};

File: LeetCode/Python/0019_Remove_Nth_Node_From_End_of_List.py
# 19. Remove Nth Node From End of List
# https://leetcode.com/problems/remove-nth-node-from-end-of-list/

from typing import Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        sNode = ListNode(0, head)
        cur = prev = sNode

        for i in range(n + 1):
            cur = cur.next

        while cur != None:
            cur = cur.next
            prev = prev.next

        prev.next = prev.next.next
        return sNode.next
